examples/download_progress/src/download.rs

```diff
@@ -1,89 +1,105 @@
-use iced_futures::futures;
+use std::hash::Hash;
 
-// Just a little utility function
-pub fn file<T: ToString>(url: T) -> iced::Subscription<Progress> {
-    iced::Subscription::from_recipe(Download {
-        url: url.to_string(),
-    })
-}
+use bytes::BufMut;
+use futures_util::stream::{self, BoxStream};
+use reqwest;
 
 pub struct Download {
-    url: String,
+    pub url: String,
+}
+
+pub enum State {
+    Ready(String),
+    Downloading {
+        url: String,
+        response: reqwest::Response,
+        total: u64,
+        bytes: Vec<u8>,
+    },
+    Finished,
 }
 
 // Make sure iced can use our download stream
 impl<H, I> iced_native::subscription::Recipe<H, I> for Download
 where
     H: std::hash::Hasher,
 {
-    type Output = Progress;
+    type Output = (String, Progress);
 
     fn hash(&self, state: &mut H) {
-        use std::hash::Hash;
-
         std::any::TypeId::of::<Self>().hash(state);
         self.url.hash(state);
     }
 
     fn stream(
         self: Box<Self>,
-        _input: futures::stream::BoxStream<'static, I>,
-    ) -> futures::stream::BoxStream<'static, Self::Output> {
-        Box::pin(futures::stream::unfold(
-            State::Ready(self.url),
+        _input: BoxStream<'static, I>,
+    ) -> BoxStream<'static, Self::Output> {
+        Box::pin(stream::unfold(
+            State::Ready(self.url.to_string()),
             |state| async move {
                 match state {
-                    State::Ready(url) => {
-                        let response = reqwest::get(&url).await;
-
-                        match response {
-                            Ok(response) => {
-                                if let Some(total) = response.content_length() {
-                                    Some((
-                                        Progress::Started,
-                                        State::Downloading {
-                                            response,
-                                            total,
-                                            downloaded: 0,
-                                        },
-                                    ))
-                                } else {
-                                    Some((Progress::Errored, State::Finished))
-                                }
-                            }
-                            Err(_) => {
-                                Some((Progress::Errored, State::Finished))
+                    State::Ready(url) => match reqwest::get(&url).await {
+                        Ok(response) => {
+                            if let Some(total) = response.content_length() {
+                                Some((
+                                    (url.to_string(), Progress::Started),
```

Comment:
We probably should be using `clone` instead of `to_string`, as it better conveys the intent.